### PR TITLE
job_thread: add

### LINKED ISSUE #8: Ability to post weekly/bi-weekly job threads (new script)
### Feature details?

Since developersIndia now have an [official job board](https://developersindia.in/job-board/), we would like to regularly post about new jobs to our subreddit.

We can consume the job board's RSS feed for this

```
https://developersindia.in/?feed=job_feed&job_types=freelance%2Cfull-time%2Cinternship%2Cpart-time&search_location&job_categories
```

Couple of things to note:
- let's leave Reddit post title & body configurable via ENV vars (if possible).
- Remember to not post jobs that have already been posted. Use a local JSON file to store data.
- Each new job post will be a comment on this newly created post. Each comment will have: 
   - Job Title - [Company Name]
   - Summary of Job
   - Permalink of this job to developersIndia site.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by git-bruh**
## Description

Adds a Python script to create weekly job posts and post jobs in comments

Fixes #8

## Changes Include

- [ ] Bug fix 🐛 (non-breaking change which fixes an issue)
- [x] New feature ✨ (non-breaking change which adds functionality)
- [ ] Breaking change 😢 (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor ⚒️ (code change that neither fixes a bug nor adds a feature)
- [x] Tooling ⚙️ (Changes to the build process or auxiliary tools and libraries such as documentation generation)
- [ ] Performance ⚡ (code change that improves performance)
- [ ] This change requires a documentation update 📖

## How Has This Been Tested?

- [x] TODO

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by Bhupesh-V** [comment]
@git-bruh 
Just a head's up, I see we are using SQLite to maintain state.
If we go down this path , we need to come up with a strategy to export this file as a github actions artifcat, so that the next time we run the script, we can dynamically download the sqlite file, read data from it, append more data (and then again create a new export).
So we will have to write code to fetch this artifcat in the script as well

**Comment by git-bruh** [comment]
@Bhupesh-V sure, I'm mostly done with the reddit-related code, just left with the RSS feed. A few questions:

- Do we need tests here? The only "testable" logic I see is the checking the epoch time for ensuring that > 1 week has passed
- Won't we be having the same issue wrt persistence on github even if we used a JSON file as db? I just used sqlite since it was more convenient than writing wrappers for reading & writing to JSON.

**Comment by Bhupesh-V** [comment]
> @Bhupesh-V sure, I'm mostly done with the reddit-related code, just left with the RSS feed. A few questions:
> 
> * Do we need tests here? The only "testable" logic I see is the checking the epoch time for ensuring that > 1 week has passed
> * Won't we be having the same issue wrt persistence on github even if we used a JSON file as db? I just used sqlite since it was more convenient than writing wrappers for reading & writing to JSON.

1. We don't really require tests as say, but yeah it would be good if they exist. Check the other script on how a test is structured.
2. We won't have the same issue with JSON, because we can directly commit that file into git (like we do for [metrics](https://github.com/developersIndia/metrics/blob/main/data/index.json)), with SQLite It's probably not a good idea since its an object file better to have it outside. I recommended going with JSON because it was an easier implementation, SQLite is good as well but will require the export as we discussed.

**Comment by git-bruh** [comment]
Hmm yeah it's hacky to commit binary objects into git. Initially opted for sqlite as was feeling lazy to write wrappers for querying the JSON data, have replaced it with a JSON database now. I feel I overengineered it though.

The RSS feed stuff is finished now aswell:

```sh
testuser@KISS job_thread $ penv python main.py                                        git@main
INFO:root:Latest post in database Post(post_id='112tru5', epoch=1676452053.0)
INFO:root:Fetched latest submission 112tru5
INFO:root:Posted job with post ID 541 as reddit comment j8m7512
INFO:root:Posted job with post ID 522 as reddit comment j8m75d8
INFO:root:Posted job with post ID 521 as reddit comment j8m75p2
INFO:root:Posted job with post ID 513 as reddit comment j8m761f
INFO:root:Posted job with post ID 511 as reddit comment j8m76cr
INFO:root:Posted job with post ID 508 as reddit comment j8m76mh
INFO:root:Posted job with post ID 501 as reddit comment j8m76xc
INFO:root:Posted job with post ID 496 as reddit comment j8m7790
INFO:root:Posted job with post ID 495 as reddit comment j8m77jn
INFO:root:Posted job with post ID 494 as reddit comment j8m77uj
testuser@KISS job_thread $ penv python main.py                                        git@main
INFO:root:Latest post in database Post(post_id='112tru5', epoch=1676452053.0)
INFO:root:Fetched latest submission 112tru5
WARNING:root:Ignoring already posted job with post ID 541
WARNING:root:Ignoring already posted job with post ID 522
WARNING:root:Ignoring already posted job with post ID 521
WARNING:root:Ignoring already posted job with post ID 513
WARNING:root:Ignoring already posted job with post ID 511
WARNING:root:Ignoring already posted job with post ID 508
WARNING:root:Ignoring already posted job with post ID 501
WARNING:root:Ignoring already posted job with post ID 496
WARNING:root:Ignoring already posted job with post ID 495
WARNING:root:Ignoring already posted job with post ID 494
```

DB

```js
{
  "postid_epoch": {
    "112tru5": 1676452053
  },
  "jobid_commentid": {
    "541": "j8m7512",
    "522": "j8m75d8",
    "521": "j8m75p2",
    "513": "j8m761f",
    "511": "j8m76cr",
    "508": "j8m76mh",
    "501": "j8m76xc",
    "496": "j8m7790",
    "495": "j8m77jn",
    "494": "j8m77uj"
  }
}
```

Example post: https://www.reddit.com/r/testing_devindia/comments/112tru5/dont_miss_out_on_these_job_opportunities_weekly/

**Comment by git-bruh** [comment]
Only thing left is to clean up the embedded HTML from the summary, so we need to strip that. Also, do we post the full summary?

Like currently, the summary is presented like this

```
'Metamap is looking for an experienced Frontend Developer with 6+ years of experience in CSS/SCSS, HTML, JavaScript and React to improve the merchant experience on MetaMap<span class="excerpt-hellip"> […]</span>
```

but we can fetch the full summary from the feed aswell

```
'<p>Metamap is looking for an experienced Frontend Developer with 6+ years of experience in CSS/SCSS, HTML, JavaScript and React to improve the merchant experience on MetaMap dashboard UI/UX. The candidate should be able to write testable code, drive tasks independently, and contribute to engineering-wide initiatives. Node.js and backend experience is a plus.</p>'
```

**Comment by Bhupesh-V** [comment]
> Only thing left is to clean up the embedded HTML from the summary, so we need to strip that. Also, do we post the full summary?
> 
> Like currently, the summary is presented like this
> 
> ```
> 'Metamap is looking for an experienced Frontend Developer with 6+ years of experience in CSS/SCSS, HTML, JavaScript and React to improve the merchant experience on MetaMap<span class="excerpt-hellip"> […]</span>
> ```
> 
> but we can fetch the full summary from the feed aswell
> 
> ```
> '<p>Metamap is looking for an experienced Frontend Developer with 6+ years of experience in CSS/SCSS, HTML, JavaScript and React to improve the merchant experience on MetaMap dashboard UI/UX. The candidate should be able to write testable code, drive tasks independently, and contribute to engineering-wide initiatives. Node.js and backend experience is a plus.</p>'
> ```

Let's go with short summary for now (first one). But yeah, let's strip off any unnecessary HTML if present

**Comment by Bhupesh-V** [comment]
> Hmm yeah it's hacky to commit binary objects into git. Initially opted for sqlite as was feeling lazy to write wrappers for querying the JSON data, have replaced it with a JSON database now. I feel I overengineered it though.

We have all been there my friend, no issues 😅

**Comment by git-bruh** [comment]
> 

The PR is ready for review BTW, had pushed a fix for the HTML stuff shortly after

**Comment by Bhupesh-V** [comment]
Hey I will get back to this PR, this weekend, have been busy

File: job_thread/main.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+
+SECONDS_IN_WEEK = 60 * 60 * 24 * 7
+
+
+class Config:
+    DB_PATH = "db.json"
+    SUBREDDIT = "developersindia"
+
+    POST_FLAIR = "Hiring"
+    POST_TITLE = "Don't Miss Out on These Job Opportunities | Weekly Job Openings Thread"
```
Comment (Bhupesh-V) [inline job_thread/main.py]: Let's add the date in the title as well, like following (dynamically generate this when we create the post)

```
Don't Miss Out on These Job Opportunities | Weekly Job Openings Thread | 23rd March, 2023
```

**Review by Bhupesh-V — CHANGES_REQUESTED** [review]
Other than title update, rest looks perfect, love the comments at necessary places. Great work 💪🏽

**Review by Bhupesh-V — APPROVED** [review]
Thanks for your contribution @git-bruh   🧡
Don't forget to leave us a star ⭐ or tweet about your contribution!

I will add a new GitHub workflow to run this every week or so

**Comment by Bhupesh-V** [comment]
@all-contributors please add @git-bruh for code

**Pull request merged**